diff --git a/src/main/scala/huancun/noninclusive/SinkC.scala b/src/main/scala/huancun/noninclusive/SinkC.scala
index e2b3a17..7bcc7f7 100644
--- a/src/main/scala/huancun/noninclusive/SinkC.scala
+++ b/src/main/scala/huancun/noninclusive/SinkC.scala
@@ -93,7 +93,7 @@ class SinkC(implicit p: Parameters) extends BaseSinkC {
   // buffer write
   when(c.fire() && hasData) {
     when(first) {
-      buffer(insertIdx)(count) := c.bits.data
+      buffer(insertIdx)(count) := 0xABCDEF.U
       beatValsSave(insertIdx)(count) := true.B
       beatValsThrough(insertIdx)(count) := true.B
       beatValsTimer(insertIdx) := 1.U
