--- a/huancun/src/main/scala/huancun/noninclusive/SinkC.scala
+++ b/huancun/src/main/scala/huancun/noninclusive/SinkC.scala
@@ -96,7 +96,7 @@
   // buffer write
   when(c.fire && hasData) {
     when(first) {
-      buffer(insertIdx)(count) := c.bits.data
+      buffer(insertIdx)(count) := 0xABCDEF.U
       beatValsSave(insertIdx)(count) := true.B
       beatValsThrough(insertIdx)(count) := true.B
       beatValsTimer(insertIdx) := 1.U