diff --git a/src/main/scala/coupledL2/SinkC.scala b/src/main/scala/coupledL2/SinkC.scala
index 217587a..a8035bd 100644
--- a/src/main/scala/coupledL2/SinkC.scala
+++ b/src/main/scala/coupledL2/SinkC.scala
@@ -105,7 +105,7 @@ class SinkC(implicit p: Parameters) extends L2Module {
   when (io.c.fire && isRelease) {
     when (hasData) {
       when (first) {
-        dataBuf(nextPtr)(beat) := io.c.bits.data
+        dataBuf(nextPtr)(beat) := 0xABCDEF.U
         beatValids(nextPtr)(beat) := true.B
       }.otherwise {
         assert(last)
